File: GIMMECore/source/Adaptation/AlgDefStructs/FitnessAlg.cpp
#include "../../../headers/Adaptation/AlgDefStructs/FitnessAlg.h"

double RandomFitness::calculate(Player* player, InteractionsProfile interactionsProfile, RegressionAlg regAlg)
{
	return 0.0;
}

double SimulationsOptimalFitness::calculate(Player* player, InteractionsProfile interactionsProfile, RegressionAlg regAlg)
{
	/*double engagement = student->getCurrState().characteristics.engagement;
		double ability = student->getCurrState().characteristics.ability;*/
	PlayerState predictedState = PlayerState(player->getCurrState());

	//((SimStudent*)student)->calcReaction(&predictedState, currIteration);
	double abilityInc = predictedState.characteristics.ability - player->getCurrState().characteristics.ability;
	return abilityInc;
}

double FundamentedFitness::calculate(Player* player, InteractionsProfile interactionsProfile, RegressionAlg regAlg)
{
	FitnessSort sort = FitnessSort(interactionsProfile);
	PlayerState predictedModel = regAlg.predict(interactionsProfile, player, &sort);
	return 0.5*(predictedModel.characteristics.ability) + 0.5*predictedModel.characteristics.engagement; //ability must be normalized to [0,1]
}

File: GIMMECore/Adaptation/AlgDefStructs/FitnessAlg.py
from abc import ABC, abstractmethod
from Adaptation.AlgDefStructs.RegressionAlg import * 

class FitnessAlg(ABC):
	@abstractmethod
	def calculate(player, interactionsProfile, regAlg):
		pass


class RandomFitness(FitnessAlg):
	def calculate(player, interactionsProfile, regAlg):
		return 0.0


class SimulationsOptimalFitness(FitnessAlg):
	def calculate(player, interactionsProfile, regAlg):
		predictedState = PlayerState(player.getCurrState());

		abilityInc = predictedState.characteristics.ability - player.getCurrState().characteristics.ability;
		return abilityInc;

class WeightedFitness(FitnessAlg):
	def __init__(self, modelWeights):
		self.modelWeights = modelWeights;

	def calculate(player, interactionsProfile, regAlg):
		predictedModel = regAlg.predict(interactionsProfile, player)
		return self.modelWeights.ability*(predictedModel.characteristics.ability) + self.modelWeights.engagement*predictedModel.characteristics.engagement; #ability must be normalized to [0,1]